### ISSUE TITLE
Job not rerun when the configmap mapped to a volume is modified

### ISSUE THREAD

**Issue by SebTardif** [fluxcd/flux#3469]
I have a job that is used to configure my service. It's a popular pattern at least in my environment, because init container is before the service started and so the service REST api and others are still not available.

My configuration is a script file, and I mount it in the job, and so with the idea of Flux, if I modify the script (configuration) and I push to my git repo, I expect my script run again.

Please advice if this should work (a bug), or Flux could be improved to handle this (enhancement), and if/or there is a good workaround.

**Comment by kingdonb** [comment]
To briefly clarify first, this repo belongs to Flux v1 which is in maintenance mode. If improvements are needed to resolve this, then it will most likely be unsolvable as this type of change in behavior could constitute a breaking change without an opt-in.

Users are recommended to upgrade to Flux v2. With that out of the way, given you might already know that, or might actually be looking for help with Flux v2 and just haven't noticed where you posted...

The problem as I understand it is that jobs (which are immutable) run only once. This doesn't seem to be configurable in Kubernetes, https://serverfault.com/a/846293/166665

It seems to me, we should expect that if the job is changed, since it is immutable, then Flux will not be able to update it, and if we are lucky then Flux is smart enough to know this rather than error, and there is a way to induce it to run again through Flux deleting and recreating it. Maybe updating an annotation (you can try) you will find the job gets recreated and run again?

I am not sure about this, I haven't really worked with Job as a free-standing (not CronJob) resource. Can you try that and see if it works?

**Comment by kingdonb** [comment]
^ above (This may or may not be a good solution for you, if it works.)

I went looking for information about this in Flux v2 repos, thinking maybe it handles Job better than Flux v1 did. So far not finding much. Here's one discussion thread that talks about creating `Jobs`: https://github.com/fluxcd/flux2/discussions/146

The discussion in that thread indicates that Jobs aren't really handled well in GitOps. At least not with Kustomize version of GitOps. (Helm on the other hand is a different story, and is made for this moment... enter Helm Controller, now is your time to shine!) so I will suggest another alternative approach is to use a Helm chart, and run your job in a Helm lifecycle hook.

There are the post-install and post-upgrade hooks:

https://helm.sh/docs/topics/charts_hooks/#the-available-hooks

**Comment by kingdonb** [comment]
Though this issue report is more nuanced, I believe this is actually a duplicate of #2440.

Closing.

**Comment by kingdonb** [comment]
To reiterate, the issue is that Flux does not work well with Job resources, you can apply them, you can delete them, you can apply them again to run them a second time, this workflow is clunky, and it hasn't been fixed to the best of my knowledge in Flux v2, either. There's no clean way to support.

I tested with Flux v2, rewriting annotations is possible on a Job resource, it doesn't do anything. Changes to the job template fail, blocking the Kustomization from validating and applying if validation is enabled.

There is [stefanprodan/kjob](/stefanprodan/kjob) which addresses this in a way by making it easy to create a job from a CronJob imperatively, but it is not in any way gitops, it requires direct cluster access in order to trigger a new job to run.

Sorry I don't have a better answer! Feel free to start a new discussion on the Flux 2 repo, or comment on one of the existing ones. There might be some improvements possible, but I'm not sure what. It seems like this turf is well explored and there isn't much to be done within the bounds of what "ideologically pure" things Flux can do while remaining strictly gitops.

**Comment by kingdonb** [comment]
I have spoken too soon, one of the maintainers alerted me to the fact that this is basically already solved in Flux v2. There is `Kustomization.spec.force` in Flux v2's Kustomize Controller which solves this.

**Comment by kingdonb** [comment]
You can set `force: true` and resources which are blocked from update in such a way.

From the Kustomize CRD spec:

```
    // Force instructs the controller to recreate resources
    // when patching fails due to an immutable field change.
    // +kubebuilder:default:=false
    // +optional
    Force bool `json:"force,omitempty"`
```

So, if you are able to use Flux v2, this issue can be solved. If you are stuck on Flux v1 and cannot upgrade, we are interested to know if there is a technical reason. All users are encouraged to migrate to Flux v2 as early as possible. Thanks for using Flux.
